src/logger/src/logger.rs

```diff
@@ -1,776 +1,343 @@
+use std::os::unix::fs::OpenOptionsExt;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::sync::{Mutex, OnceLock};
+use std::thread;
+
+use log::{Log, Metadata, Record};
+use serde::{Deserialize, Serialize};
 use utils::time::LocalTime;
 
-use super::extract_guard;
-use crate::init;
-use crate::init::Init;
 use crate::metrics::{IncMetric, METRICS};
 
-/// Type for returning functions outcome.
-pub type Result<T> = result::Result<T, LoggerError>;
-
-// Values used by the Logger.
-const DEFAULT_MAX_LEVEL: LevelFilter = LevelFilter::Warn;
-
-lazy_static! {
-    static ref _LOGGER_INNER: Logger = Logger::new();
-
-    /// Static instance used for handling human-readable logs.
-    pub static ref LOGGER: &'static Logger = {
-        set_logger(_LOGGER_INNER.deref()).expect("Failed to set logger");
-        _LOGGER_INNER.deref()
-    };
-}
-
-/// Logger representing the logging subsystem.
-// All member fields have types which are Sync, and exhibit interior mutability, so
-// we can call logging operations using a non-mut static global variable.
-pub struct Logger {
-    init: Init,
-    // Human readable logs will be outputted here.
-    log_buf: Mutex<Box<dyn Write + Send>>,
-    show_level: AtomicBool,
-    show_file_path: AtomicBool,
-    show_line_numbers: AtomicBool,
-    instance_id: RwLock<String>,
-}
-
-impl fmt::Debug for Logger {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Logger")
-            .field("init", &self.init)
-            .field("log_buf", &"?")
-            .field("show_level", &self.show_level)
-            .field("show_file_path", &self.show_file_path)
-            .field("show_line_numbers", &self.show_line_numbers)
-            .field("instance_id", &self.instance_id)
-            .finish()
-    }
-}
+/// Default instance id.
+pub const DEFAULT_INSTANCE_ID: &str = "anonymous-instance";
+/// Instance id.
+pub static INSTANCE_ID: OnceLock<String> = OnceLock::new();
+
+/// The logger.
+///
+/// Default values matching the swagger specification (`src/api_server/swagger/firecracker.yaml`).
+pub static LOGGER: Logger = Logger(Mutex::new(LoggerConfiguration {
+    target: None,
+    filter: LogFiler {
+        level: log::LevelFilter::Warn,
+    },
+    format: LogFormat {
+        show_level: false,
+        show_log_origin: false,
+    },
+}));
+
+/// Error type for [`Logger::update`].
+#[derive(Debug, thiserror::Error)]
+#[error("Failed to open target file: {0}")]
+pub struct LoggerUpdateError(pub std::io::Error);
 
 impl Logger {
-    /// Creates a new instance of the current logger.
-    fn new() -> Logger {
-        Logger {
-            init: Init::new(),
-            log_buf: Mutex::new(Box::new(sink())),
-            show_level: AtomicBool::new(true),
-            show_line_numbers: AtomicBool::new(true),
-            show_file_path: AtomicBool::new(true),
-            instance_id: RwLock::new(String::new()),
+    /// Applies the given logger configuration the logger.
```

Comment:
This function has significantly less documentation than the functions is replaces. I think we should try to preserve the parts of the documentation that are still relevant.

Reply:
Are there specific parts of the documentation you think are worth keeping?

Reply:
Yeah, looking at it outside of GitHub's diff view I agree, the code is simpler with a lot less functions now, so this is fine by me :)